File: taurus-protect-sdk-python/taurus_protect/helpers/signature_verifier.py
"""Signature verification utilities for governance rules."""

from __future__ import annotations

import base64
import binascii
import hashlib
from typing import List, Optional, Sequence

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey

from taurus_protect._strict_base64 import strict_b64decode
from taurus_protect.crypto.signing import verify_signature
from taurus_protect.errors import IntegrityError
from taurus_protect.models.governance_rules import GovernanceRules, RuleUserSignature


def verify_governance_rules(
    rules: GovernanceRules,
    min_valid_signatures: int,
    super_admin_keys: List[EllipticCurvePublicKey],
) -> None:
    """
    Verify that governance rules are signed by enough DISTINCT SuperAdmin keys.

    min_valid_signatures counts distinct signing keys, not signature entries: ECDSA
    is randomized, so counting entries would let a single key produce as many valid
    signatures as any threshold demands, making 2-of-N no stronger than 1-of-N.

    Args:
        rules: The governance rules to verify.
        min_valid_signatures: Minimum number of distinct valid signers required.
        super_admin_keys: List of SuperAdmin public keys for verification.

    Raises:
        IntegrityError: If verification fails or not enough valid signatures.
        ValueError: If arguments are invalid.

    Example:
        >>> verify_governance_rules(rules, min_valid_signatures=2, super_admin_keys=keys)
    """
    if rules is None:
        raise ValueError("rules cannot be None")
    if min_valid_signatures <= 0:
        raise ValueError("min_valid_signatures must be positive")
    if not super_admin_keys:
        raise ValueError("super_admin_keys cannot be empty")

    if rules.rules_container is None:
        raise IntegrityError("Governance rules verification failed: rulesContainer is null")

    signatures = rules.rules_signatures
    if not signatures:
        raise IntegrityError("Governance rules verification failed: no signatures present")

    # Decode the rules container
    try:
        rules_data = strict_b64decode(rules.rules_container)
    except (binascii.Error, ValueError) as e:
        raise IntegrityError(f"Governance rules verification failed: invalid base64 encoding: {e}") from e

    try:
        verify_governance_rules_signatures(
            rules_data, signatures, super_admin_keys, min_valid_signatures
        )
    except IntegrityError as e:
        raise IntegrityError(f"Governance rules verification failed: {e}") from e


def verify_governance_rules_signatures(
    rules_container_data: bytes,
    signatures: Sequence[RuleUserSignature],
    super_admin_keys: List[EllipticCurvePublicKey],
    min_valid_signatures: int,
) -> None:
    """
    Verify that the rules container is signed by enough DISTINCT SuperAdmin keys.

    This is the single place the threshold is evaluated -- every caller routes here.
    min_valid_signatures counts distinct signing keys, not signature entries: ECDSA
    is randomized, so counting entries would let a single key produce as many valid
    signatures as any threshold demands, making 2-of-N no stronger than 1-of-N.

    Args:
        rules_container_data: The raw signed rules container bytes.
        signatures: The signature entries to check.
        super_admin_keys: List of SuperAdmin public keys for verification.
        min_valid_signatures: Minimum number of distinct valid signers required.

    Raises:
        IntegrityError: If too few distinct SuperAdmin keys signed.
        ValueError: If min_valid_signatures is not positive.
    """
    if min_valid_signatures <= 0:
        raise ValueError("min_valid_signatures must be positive")
    if not rules_container_data:
        raise IntegrityError("rules container data cannot be empty")
    if not super_admin_keys:
        raise IntegrityError("no SuperAdmin keys configured for verification")
    if not signatures:
        raise IntegrityError("no signatures provided")

    signers = set()
    for sig in signatures:
        if not sig.signature:
            continue
        key = _matching_key(rules_container_data, sig.signature, super_admin_keys)
        if key is not None:
            signers.add(key_fingerprint(key))

    if len(signers) < min_valid_signatures:
        raise IntegrityError(
            f"only {len(signers)} distinct valid signers found, "
            f"minimum {min_valid_signatures} required"
        )


def _matching_key(
    data: bytes,
    signature_b64: str,
    super_admin_keys: List[EllipticCurvePublicKey],
) -> Optional[EllipticCurvePublicKey]:
    """Return the first configured key that verifies the signature, else None."""
    for public_key in super_admin_keys:
        try:
            if verify_signature(public_key, data, signature_b64):
                return public_key
        except (InvalidSignature, ValueError):
            # Signature verification failed for this key, try next
            # InvalidSignature: cryptographic verification failed
            # ValueError: malformed signature or key format
            continue
    return None


def key_fingerprint(public_key: EllipticCurvePublicKey) -> bytes:
    """Identify a key by its encoded bytes, so the same key configured twice counts once."""
    der = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return hashlib.sha256(der).digest()


def is_valid_signature(
    data: bytes,
    signature_b64: str,
    super_admin_keys: List[EllipticCurvePublicKey],
) -> bool:
    """
    Verify a signature against the provided SuperAdmin public keys.

    Tries each key in sequence and returns True if any key validates
    the signature.

    Args:
        data: The data that was signed.
        signature_b64: Base64-encoded signature.
        super_admin_keys: List of SuperAdmin public keys to try.

    Returns:
        True if the signature is valid for any of the keys.
    """
    return _matching_key(data, signature_b64, super_admin_keys) is not None


def verify_raw_signature(
    data: bytes,
    signature: bytes,
    public_key: EllipticCurvePublicKey,
) -> bool:
    """
    Verify a raw signature against a single public key.

    Args:
        data: The data that was signed.
        signature: The raw signature bytes (not base64).
        public_key: The public key to verify against.

    Returns:
        True if the signature is valid.
    """
    if data is None or signature is None or public_key is None:
        return False

    # Convert raw bytes to base64 for verify_signature
    signature_b64 = base64.b64encode(signature).decode("utf-8")
    return verify_signature(public_key, data, signature_b64)

File: taurus-protect-sdk-typescript/src/helpers/signature-verifier.ts
/**
 * Signature verification utilities for governance rules.
 *
 * This module provides functions to verify SuperAdmin signatures on
 * governance rules and other cryptographically protected data.
 */

import { createHash, type KeyObject } from "crypto";

import { verifySignature, decodePublicKeysPem } from "../crypto";
import { ConfigurationError, IntegrityError } from "../errors";
import { isCryptoVerificationError } from "../crypto";
import { strictBase64Decode } from "./strict-base64";

/**
 * Returns the first key that verifies the signature, or undefined.
 */
function matchingKey(
  data: Buffer | string,
  signatureBase64: string,
  publicKeys: KeyObject[]
): KeyObject | undefined {
  const dataBuffer =
    typeof data === "string" ? Buffer.from(data, "utf-8") : data;

  for (const publicKey of publicKeys) {
    try {
      if (verifySignature(publicKey, dataBuffer, signatureBase64)) {
        return publicKey;
      }
    } catch (error: unknown) {
      // This key did not verify the signature; try the next configured key.
      if (isCryptoVerificationError(error)) {
        continue;
      }
      throw error;
    }
  }

  return undefined;
}

/**
 * Identifies a key by its encoded bytes, so the same key configured twice
 * counts as one signer.
 */
export function keyFingerprint(publicKey: KeyObject): string {
  const der = publicKey.export({ type: "spki", format: "der" });
  return createHash("sha256").update(der).digest("hex");
}

/**
 * Verifies a single signature against a set of public keys.
 *
 * @param data - The signed data (as buffer or string)
 * @param signatureBase64 - Base64-encoded signature
 * @param publicKeys - Array of public keys to verify against
 * @returns true if the signature is valid for any of the keys
 */
export function isValidSignature(
  data: Buffer | string,
  signatureBase64: string,
  publicKeys: KeyObject[]
): boolean {
  return matchingKey(data, signatureBase64, publicKeys) !== undefined;
}

/**
 * Verifies a base64 signature against ONE public key.
 *
 * The single-key form the other three SDKs expose (Go VerifySignatureWithKey, Java
 * SignatureVerifier.verifySignature, Python verify_raw_signature). Without it a caller
 * checking a known signer had to pass a one-element array to
 * {@link isValidSignature}, which reads as a quorum check.
 *
 * Returns false rather than throwing for a missing key or malformed signature: the data,
 * signature and key are all public here, so there is nothing to leak and nothing to
 * distinguish between failure modes.
 *
 * @param data - The signed payload
 * @param signatureBase64 - The signature, base64-encoded raw r||s
 * @param publicKey - The public key to check against
 * @returns true when the signature verifies under this key
 */
export function verifySignatureWithKey(
  data: Buffer | string,
  signatureBase64: string,
  publicKey: KeyObject | undefined | null
): boolean {
  if (!publicKey) {
    return false;
  }
  const dataBuffer = typeof data === "string" ? Buffer.from(data, "utf-8") : data;
  return verifySignature(publicKey, dataBuffer, signatureBase64);
}

/**
 * Verifies that the rules container is signed by the required number of DISTINCT
 * SuperAdmin keys. Throws on failure, returns normally on success.
 *
 * This is the single place the threshold is evaluated — every caller routes here.
 * minValidSignatures counts distinct signing keys, not signature entries: ECDSA is
 * randomized, so counting entries would let a single key produce as many valid
 * signatures as any threshold demands, making a 2-of-N quorum no stronger than 1-of-N.
 *
 * @throws {@link ConfigurationError} If minValidSignatures is not positive
 * @throws {@link IntegrityError} If too few distinct SuperAdmin keys signed
 */
export function verifyGovernanceRulesSignatures(
  rulesContainerData: Buffer,
  signatures: ReadonlyArray<{ readonly signature?: string }>,
  superAdminKeys: KeyObject[],
  minValidSignatures: number
): void {
  if (minValidSignatures <= 0) {
    throw new ConfigurationError("minValidSignatures must be positive");
  }

  if (rulesContainerData.length === 0) {
    throw new IntegrityError("rules container data cannot be empty");
  }

  if (superAdminKeys.length === 0) {
    throw new IntegrityError("no SuperAdmin keys configured for verification");
  }

  if (signatures.length === 0) {
    throw new IntegrityError("no signatures provided");
  }

  const signers = new Set<string>();

  for (const sig of signatures) {
    if (!sig.signature) continue;

    const key = matchingKey(rulesContainerData, sig.signature, superAdminKeys);
    if (key) {
      signers.add(keyFingerprint(key));
    }
  }

  if (signers.size < minValidSignatures) {
    throw new IntegrityError(
      `insufficient distinct valid SuperAdmin signers: got ${signers.size}, need ${minValidSignatures}`
    );
  }
}

/**
 * Verifies governance rules signatures using SuperAdmin keys in PEM form.
 *
 * @param rulesContainerBase64 - Base64-encoded rules container
 * @param signaturesBase64 - Base64-encoded signatures array
 * @param minValidSignatures - Minimum number of DISTINCT valid signers required
 * @param superAdminKeysPem - Array of SuperAdmin public keys in PEM format
 * @returns true if enough distinct signers are present
 * @throws {@link ConfigurationError} If minValidSignatures is not positive
 */
export function verifyGovernanceRules(
  rulesContainerBase64: string,
  signaturesBase64: string,
  minValidSignatures: number,
  superAdminKeysPem: string[]
): boolean {
  if (minValidSignatures <= 0) {
    throw new ConfigurationError("minValidSignatures must be positive");
  }

  if (!rulesContainerBase64 || !signaturesBase64) {
    return false;
  }

  if (superAdminKeysPem.length === 0) {
    return false;
  }

  // Decode public keys
  let publicKeys: KeyObject[];
  try {
    publicKeys = decodePublicKeysPem(superAdminKeysPem);
  } catch (error: unknown) {
    if (isCryptoVerificationError(error)) {
      return false;
    }
    throw error;
  }

  // Decode signatures from base64
  let signatures: Array<{ userId?: string; signature?: string }>;
  try {
    const decoded = Buffer.from(signaturesBase64, "base64").toString("utf-8");
    const data = JSON.parse(decoded) as unknown;
    if (Array.isArray(data)) {
      signatures = data as Array<{ userId?: string; signature?: string }>;
    } else if (data && typeof data === "object" && "signatures" in data) {
      const dataObj = data as { signatures?: unknown };
      if (!Array.isArray(dataObj.signatures)) {
        // Invalid format: signatures property exists but is not an array
        return false;
      }
      signatures = dataObj.signatures as Array<{
        userId?: string;
        signature?: string;
      }>;
    } else {
      // Invalid format: not an array and no signatures property
      return false;
    }
  } catch (error: unknown) {
    if (error instanceof SyntaxError ||
        (error instanceof Error &&
         (error.message.includes('decode') ||
          error.message.includes('JSON') ||
          error.message.includes('base64')))) {
      return false;
    }
    throw error;
  }

  // The signed data is the raw rules container bytes
  const rulesData = strictBase64Decode(rulesContainerBase64);

  try {
    verifyGovernanceRulesSignatures(
      rulesData,
      signatures,
      publicKeys,
      minValidSignatures
    );
    return true;
  } catch (error: unknown) {
    if (error instanceof IntegrityError) {
      return false;
    }
    throw error;
  }
}
